students/kegan/session07/html_render.py

```diff
@@ -0,0 +1,186 @@
+"""
+Kathryn Egan
+
+"""
+
+
+class Element:
+    """ Stores behavior and data for an Element of HTML. """
+    indent = '    '
+
+    def __init__(self, content=None, **kwargs):
+        """ Initializes an Element with content, if given,
+        and any HTML attributes of the Element.
+        Args:
+            content (str or Element) : text or a nested Element
+            kwargs (dic str:str) : HTML attributes for Element
+        """
+        if content is None:
+            self.content = []
+        else:
+            self.content = [content]
+        self.attrs = kwargs
+
+    def append(self, content):
+        """ Adds content to this Element.
+        Args:
+            content (str or Element) : text or a nested Element
+        """
+        self.content.append(content)
+
+    def render(self, file_out, ind=0):
+        """ Renders the contents of this Element into HTML
+        format and pretty prints to file using given file handler.
+        Args:
+            file_out (TextIOWrapper) : file handler
+            ind (str) : level of indentation for this Element
+        """
+        file_out.write(self.get_tag(ind))
+        file_out.write('\n')
+        for c in self.content:
+            try:
+                c.render(file_out, ind + 1)
+            except AttributeError:
+                file_out.write('{}{}\n'.format((ind + 1) * self.indent, c))
+        file_out.write(self.get_tag(ind, style='close'))
+        file_out.write('\n')
+
+    def get_tag(self, ind, style='open'):
+        """ Generates an HTML tag according to given style
+        with given level of indent. Style can be 'open',
+        'close', or 'selfclosing' (defaults to 'open').
+        Args:
+            ind (str) : level of indent
+            style (str) : style to use
+        Returns:
+            tag : formatted tag
+        """
+        substrings = {
```

Comment:
I'm not sure putting it all in a dict, just so you can pass it in to format it worth it.

though it is kind of elegant.

Reply:
everything on the html_render is ok.   Learned lots from your solution